stan/math/prim/meta/plain_type.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef STAN_MATH_PRIM_META_PLAIN_TYPE_HPP
+#define STAN_MATH_PRIM_META_PLAIN_TYPE_HPP
+
+#include <stan/math/prim/meta/plain_type.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * Determines plain (non expression) type associated with \c T. For non \c Eigen
+ * types it is the input type.
+ * @tparam T type to determine plain type of
+ */
+template <typename T, typename Enable = void>
+struct plain_type {
+  using type = std::decay_t<T>;
+};
+
+template <typename T>
+using plain_type_t = typename plain_type<T>::type;
+
+/**
+ * Determines return type of calling .eval() on eigen expression. This is the
+ * same as \c plain_type, except if input is already a plain type. In such case
+ * eval return type is a const reference to plain type.
+ * @tparam T type to determine eval return type of
+ */
```

Comment:
I think I'd phrase this as

```
/**
 * Determines type for a call to .eval() on an eigen expression. if decaying \c T gives the same type as
 * \c plain_type_t<T> then the member \c type will be \c const \c T&. 
 * Otherwise \c type will be the same as the \c type of plain_type_t<T>
 * @tparam T type to determine eval return type of
 */
```

Reply:
I find your version a bit more confusing. If you want I can add in code fragments from your version, but I would prefere not to use it in whole.

Reply:
done